src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/FlatArrayIntervalTree`1.cs

```diff
@@ -0,0 +1,430 @@
+                array[currentNodeIndex] = new Node(currentNode.Value, maxEndNodeIndex);
+                return leftMaxEndValue;
+            }
+            else
+            {
+                Contract.ThrowIfFalse(rightMaxEndValue > leftMaxEndValue);
+
+                // Set this node's max end to be the right subtree's max end.
+                var maxEndNodeIndex = array[rightChildIndex].MaxEndNodeIndex;
+                array[currentNodeIndex] = new Node(currentNode.Value, maxEndNodeIndex);
+                return rightMaxEndValue;
+            }
+        }
+    }
+
+    private static int GetLeftChildIndex(int nodeIndex)
+        => (2 * nodeIndex) + 1;
+
+    private static int GetRightChildIndex(int nodeIndex)
+        => (2 * nodeIndex) + 2;
+
+    private static int GetEnd<TIntrospector>(T value, in TIntrospector introspector)
+        where TIntrospector : struct, IIntervalIntrospector<T>
+        => introspector.GetSpan(value).End;
+
+    bool IIntervalTree<T>.Any<TIntrospector>(int start, int length, TestInterval<T, TIntrospector> testInterval, in TIntrospector introspector)
+    {
+        // Inlined version of FillWithIntervalsThatMatch, optimized to do less work and stop once it finds a match.
+        var array = _array;
+        if (array.Length == 0)
+            return false;
+
+        using var _ = s_nodeIndexPool.GetPooledObject(out var candidates);
+
+        var end = start + length;
+
+        candidates.Push(0);
+
+        while (candidates.TryPop(out var currentNodeIndex))
+        {
+            // Check the nodes as we go down.  That way we can stop immediately when we find something that matches,
+            // instead of having to do an entire in-order walk, which might end up hitting a lot of nodes we don't care
+            // about and placing a lot into the stack.
+            var node = array[currentNodeIndex];
+            if (testInterval(node.Value, start, length, in introspector))
+                return true;
+
+            if (ShouldExamineRight(array, start, end, currentNodeIndex, in introspector, out var rightIndex))
+                candidates.Push(rightIndex);
+
+            if (ShouldExamineLeft(array, start, currentNodeIndex, in introspector, out var leftIndex))
+                candidates.Push(leftIndex);
+        }
+
+        return false;
+    }
+
+    int IIntervalTree<T>.FillWithIntervalsThatMatch<TIntrospector>(
+        int start, int length, TestInterval<T, TIntrospector> testInterval,
+        ref TemporaryArray<T> builder, in TIntrospector introspector,
+        bool stopAfterFirst)
+    {
+        var array = _array;
+        if (array.Length == 0)
+            return 0;
+
+        using var _ = s_stackPool.GetPooledObject(out var candidates);
+
+        var matches = 0;
+        var end = start + length;
+
+        candidates.Push((nodeIndex: 0, firstTime: true));
+
+        while (candidates.TryPop(out var currentTuple))
```

Comment:
>while (candidates.TryPop(out var currentTuple)) [](http://example.com/codeflow?start=8&length=47)

From our offline conversation, might be nice if this and GetEnumerator shared code